src/server/gc_worker/compaction_filter.rs

```diff
@@ -154,133 +204,269 @@ struct WriteCompactionFilter {
+        if !self.remove_older {
+            match write.write_type {
+                WriteType::Rollback | WriteType::Lock => {
+                    self.mvcc_rollback_and_locks += 1;
+                    filtered = true;
+                }
+                WriteType::Put => self.remove_older = true,
+                WriteType::Delete => {
+                    self.remove_older = true;
+                    if self.is_bottommost_level {
+                        filtered = true;
+                        self.deleting_filtered = Some((key.to_vec(), Vec::new()));
+                    }
+                }
+            }
+        } else if let Some(deleting_filtered) = self.deleting_filtered.as_mut() {
+            deleting_filtered.1.push((commit_ts, sequence));
+        }
+
+        if !filtered {
+            return Ok(CompactionFilterDecision::Keep);
+        }
+        self.filtered += 1;
+        self.handle_filtered_write(write)?;
+        let decision = if self.remove_older {
+            debug_assert!(commit_ts > 0);
+            let prefix = Key::from_encoded_slice(key_prefix);
+            let skip_until = prefix.append_ts((commit_ts - 1).into()).into_encoded();
+            CompactionFilterDecision::RemoveAndSkipUntil(skip_until)
+        } else {
+            CompactionFilterDecision::Remove
+        };
+        Ok(decision)
+    }
+
+    fn meet_a_key_in_filter(&mut self) {
+        if self.is_bottommost_level && self.deleting_filtered.is_none() {
+            // The compaction hasn't met the next MVCC delete mark.
+            self.near_seek_distance += 1;
         }
     }
 
     // Do gc before a delete mark.
-    fn handle_delete_mark(&mut self, mark: &[u8]) {
-        let mut valid = self.write_iter_seek_to(mark);
+    fn handle_delete_mark(&mut self) -> Result<(), String> {
+        let (mark, mut filtered) = match self.deleting_filtered.take() {
+            Some((x, y)) => (x, y.into_iter().peekable()),
+            None => return Ok(()),
+        };
+        debug_assert!(self.is_bottommost_level);
+
+        let mut valid = self.write_iter_seek_to(&mark)?;
         let mut write_iter = self.write_iter.take().unwrap();
         if valid && write_iter.key() == mark {
             // The delete mark itself is handled in `filter`.
-            valid = write_iter.next().unwrap();
+            valid = write_iter.next()?;
         }
 
         while valid {
             let (key, value) = (write_iter.key(), write_iter.value());
-            if truncate_ts(key) != self.key_prefix.as_slice() {
+            if truncate_ts(key)? != self.key_prefix.as_slice() {
                 break;
             }
-            self.handle_filtered_write(parse_write(value));
-            self.delete_write_key(key);
-            valid = write_iter.next().unwrap();
+            let (_, commit_ts) = split_ts(key)?;
+            let mut need_delete = true;
+            while let Some((ts, s)) = filtered.peek() {
+                match commit_ts.cmp(ts) {
```

Comment:
```suggestion
                match current_ts.cmp(filtered_ts) {
```
for a better naming?